src/Featurizers/Components/WindowFeaturizerBase.h

```diff
@@ -0,0 +1,244 @@
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::pointer, T*>::value, "Make sure standard iterator handles iterator traits correctly!");
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::reference, T&>::value, "Make sure standard iterator handles iterator traits correctly!");
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::iterator_category, std::forward_iterator_tag>::value, "Make sure standard iterator handles iterator traits correctly!");
+
+        // ----------------------------------------------------------------------
+        // |  Private Data
+
+        // Pointer to the first element of the underlying data structure.
+        T*                                      _itr;
+
+        // The number of elements that the underlying data structure has.
+        size_t                                  _size;
+
+        // The index this iterator is pointer it in the underlying data structure.
+        size_t                                  _cur_index;
+
+        // The maximum number of times this iterator can be incremented. Since this iterator will
+        // wrap at the end of the data structure, there is no real "end" of the data so this is
+        // used to determine when we have met the "end".
+        size_t                                  _max_increments;
+
+        // How many times we have incremented this iterator.
+        size_t                                  _cur_increment;
+
+    public:
+        // ----------------------------------------------------------------------
+        // |  Public Methods
+
+        // Default constructor, already at the "end".
+        CircularIterator()
+            : _itr(nullptr), _size(0), _cur_index(0),
+            _max_increments(0), _cur_increment(0) {
+        }
+
+        CircularIterator(T* T, size_t container_max_size, size_t max_increments, size_t starting_offset = 0)
+            : _itr(T), _size(container_max_size), _cur_index(starting_offset),
+            _max_increments(max_increments), _cur_increment(0) {
+        }
+
+        // Pre-increment operator.
+        CircularIterator& operator++ () {
+            assert(++_cur_increment <= _max_increments);
+
+            ++_cur_index;
+            _cur_index %= _size;
+            return *this;
+        }
+
+        // Post-increment operator.
+        CircularIterator operator++ (int) {
+            CircularIterator tmp(*this);
+
+            assert(++_cur_increment <= _max_increments);
+
+            ++_cur_index;
+            _cur_index %= _size;
+            return tmp;
+        }
+
+        // Two-way comparison: v.begin() == v.cbegin() and vice versa.
+        template<class OtherType>
+        bool operator == (const CircularIterator<OtherType>& rhs) const {
```

Comment:
>bool operator == (const CircularIterator<OtherType>& rhs) const { [](start = 8, length = 65)

There is a subtle bug here in that a very old iterator could compare as equal to a new iterator after the internal memory buffer has wrapped. I would recommend adding a version value that is set in the iterator's ctor and checking that as well. In the CircularBuffer class, update that version value every time an element is added.

Reply:
Can you explain more on this? I am not sure I follow the situation.

Reply:
Resolving per offline discussions. Following STL templates and old iterators are invalid when the underlying data is changed.